### ISSUE TITLE
Stretchy header effect

### ISSUE THREAD

**Issue by maxlapides** [flutter/flutter#27373]
I'm looking to create a header effect like this:

![stretchy header](https://raw.githubusercontent.com/gskbyte/GSKStretchyHeaderView/master/screenshots/stretchy_blur_small.gif)

Specifically, I want it so when the user overscrolls on iOS, the image should zoom to fill the extra space. Then, on Android and iOS, when the user starts scrolling down the screen, the header should be scrolled away with a parallax effect.

I pulled the above screen video from this package here: https://github.com/gskbyte/GSKStretchyHeaderView

Another example can be found here: https://medium.freecodecamp.org/tutorial-creating-stretchy-layouts-on-ios-using-auto-layout-3fa974fa5e28

I was playing around with `SliverPersistentHeader`s and `FlexibleSpaceBar`s but I was unable to get this effect quite how I wanted it.

**Comment by zoechi** [comment]
Please consider asking support questions in one of the other channels listed at http://flutter.io/support .

**Comment by maxlapides** [comment]
I posted the question on StackOverflow: https://stackoverflow.com/questions/54580218/stretchy-header-effect

This does feel like the kind of thing that Flutter would want to officially support in some way, though, since it's a pretty common UI pattern.

**Comment by zoechi** [comment]
Flutter is about lightweight widgets that can be composed to more powerful widgets.
What you propose would probably require lots of options to customize for individual needs and therefore probably not a good fit for the framework.

**Comment by maxlapides** [comment]
That makes sense! I wonder if there's some options or sliver widgets that Flutter could provide that would make this kind of thing easier, though. Or if I'm just missing something that already exists!

**Comment by kangwang1988** [comment]
@maxlapides 
See flutter_gallery/Studies/Pesto for more.
![image](https://user-images.githubusercontent.com/817851/52465208-c45e4500-2bb8-11e9-9ae0-5fe05b6e8874.png)

**Comment by maxlapides** [comment]
@kangwang1988 thanks for the tip! The `SliverAppBar` demo is handy. However, it doesn't demonstrate how I can get the stretchy background effect that I'm looking for. I'm still stuck on this one, unfortunately.

**Comment by zoechi** [comment]
@maxlapides It would be better to ask on StackOverflow and only create an issue if it turns out to be a bug or a missing feature that prevents you from doing what you want.

**Comment by maxlapides** [comment]
@zoechi I did create an issue on StackOverflow, but so far I haven't found an answer. As far as I can tell, this is a missing feature that is preventing me from doing what I want.

**Comment by zoechi** [comment]
I'll reopen and let the widget/framework maintainer decide.

**Comment by HiroshiHorie** [comment]
I'm trying to achieve the same effect without success. This should be a SliverAppBar option since it is such a common UI pattern in iOS. Maybe with LayoutBuilder it is possible?

**Comment by ninnepinne** [comment]
@zoechi 

> Flutter is about lightweight widgets that can be composed to more powerful widgets.
What you propose would probably require lots of options to customise for individual needs and therefore probably not a good fit for the framework.

I agree with your statement, but Flutter is also very much a multi platform tool.
And as a developer you kind of expect that eg. SliverAppBar and SliverPersistentHeader would work on iOS (and behave iOS-like).
It's a common pattern on iOS too.

Right now you are very limited regarding any type of expanding header on iOS.

I also hit this bug that is somewhat related (Bouncing scroll physics breaks sliver header):
https://github.com/flutter/flutter/issues/29264

To sum it up: Flutter could support sliver headers better for iOS.

**Comment by sky1095** [comment]
there is a plugin available: https://pub.dev/packages/stretchy_header

**Comment by sooxiaotong** [comment]
Nice plugin but for me I need a dynamic height. And the plugin is abit to heavy with stack, my current idea is to use cliprect and scale together.

**Comment by Piinks** [comment]
This looks like the inverted behavior of SliverFillRemaining(fillOverscroll: true, ...)
I think it could work in a similar way.

**Comment by Piinks** [comment]
Hi all,
Just clarifying - what is needed here is the option for a stretch effect with `SliverAppBar`, along with a callback for when it is scrolled to a specified triggering distance? 

I am assuming this will follow a similar convention to that of `CupertinoSliverRefreshControl`.

Also, currently `SliverAppBar` has options for other behaviors as well, including **floating**, **pinning** and **snapping**. I am also assuming these should all be available in all of the possible combinations?

Feedback always welcomed. 🙂

**Comment by maxlapides** [comment]
@Piinks that sounds exactly right!

**Comment by Piinks** [comment]
Thanks for the feedback @maxlapides! 
Would you be able to tell me what the package [stretchy header](https://pub.dev/packages/stretchy_header), is lacking in meeting this need?

**Comment by Piinks** [comment]
> Nice plugin but for me I need a dynamic height. And the plugin is abit to heavy with stack, my current idea is to use cliprect and scale together.

@sooxiaotong, could you tell me what you mean by dynamic height?

**Comment by sooxiaotong** [comment]
@Piinks the header height is not dynamic. It should follow the image ratio after loaded without setting the fix height.

**Comment by Piinks** [comment]
Thanks for clarifying!

**Comment by Piinks** [comment]
#42250 will provide this feature. 
That PR has a couple of animations and sample code to illustrate the use cases. If you would like to check it out, I would love to know what you think. :)

**Comment by sooxiaotong** [comment]
![ezgif-4-2f4d2ee159c4 (1)](https://user-images.githubusercontent.com/13378059/66593086-e1746900-ebc8-11e9-830b-1bc2f2cae5f1.gif)

Hi @Piinks, 
Is it possible to do a pull to refresh like this with the indicator that fades in gradually while pulling down, because seems like the callback doesn't have the stretched percentage and stretch threshold settings. And also does the `expandedHeight` supports dynamic height base on the image?

**Comment by Piinks** [comment]
Hey @sooxiaotong !

I believe it is possible to do with #42250. Using a LayoutBuilder in the flexibleSpace of the SliverAppBar, the constraints.maxHeight will provide you with the current size of the app bar as it stretches. The percentage can be easily calculated based on your chosen 'expandedHeight`.  The stretch threshold is set to an overscroll offset of 100.0 by default, but you can customize that as well to whatever you like.

As to your other question - do you mean dynamic height as in, the app bar's expandedHeight will change to whatever height the image is? If that's the case, I do not think that is currently supported, and sounds like a separate feature. I would recommend filing a separate issue for that feature if that is what you are looking for. :)

**Comment by sooxiaotong** [comment]
@Piinks Okay thanks, looks like you have answered all my question~ I can't wait to use this pull in app now :3

**Comment by maxlapides** [comment]
Ship it 😍

**Comment by sooxiaotong** [comment]
@Piinks One last question, is it possible to use `stretchModes` with `LayoutBuilder` inside `flexibleSpace`?

**Comment by Piinks** [comment]
The proposed `StretchModes` are only used as flags for the `FlexibleSpaceBar` class. If you look how they are used in the build method of `FlexibleSpaceBar`, you can see that the effects they trigger are fairly simple, using only a few lines of code. :)

**Comment by 1111mp** [comment]
@Piinks First of all, thank you very much for your work.
Excuse me, I can only use it on the master branch now, Has it been supported on the stable channel?

**Comment by 1111mp** [comment]
@Piinks 
How to use it in the SliverPersistentHeader class and has supported it?

![屏幕录制2019-12-04下午4 44 51 2019-12-04 16_49_03](https://user-images.githubusercontent.com/31227919/70127465-0e6b6680-16b6-11ea-9f5b-fefed8743776.gif)
I want to do this.This is my current code.
```
import 'dart:ui';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class AccountPage extends StatefulWidget {
  @override
  _AccountPageState createState() => _AccountPageState();
}

class _AccountPageState extends State<AccountPage>
    with TickerProviderStateMixin {
  double extraPicHeight;
  double extraBtmHeight;
  ScrollPhysics _physics;
  double prevDy;

  bool callNow;

  AnimationController animationController;
  Animation<double> animate;

  ScrollController _scrollController = new ScrollController();

  @override
  void initState() {
    super.initState();
    extraPicHeight = 0.0;
    extraBtmHeight = 0.0;
    prevDy = 0.0;
    callNow = true;
    _physics = AlwaysScrollableScrollPhysics();
    animationController = AnimationController(
      vsync: this,
      duration: Duration(milliseconds: 300),
    );
    animate = Tween(begin: 0.0, end: 0.0).animate(animationController);
  }

  void updatePicHeight(changed) {
    extraPicHeight = changed / 3;

    setState(() {
      extraPicHeight = extraPicHeight;
    });
  }

  void runAnimate() {
    setState(() {
      animate =
          Tween(begin: extraPicHeight, end: 0.0).animate(animationController)
            ..addListener(() {
              setState(() {
                extraPicHeight = animate.value;
              });
            });
      // prevDy = 0.0;
      extraPicHeight = 0.0;
    });
  }

  // 节流
  void debounce(Function fn) {
    if (callNow) {
      fn();
      callNow = false;
    }
  }

  @override
  Widget build(BuildContext context) {
    return AnnotatedRegion<SystemUiOverlayStyle>(
      value: SystemUiOverlayStyle.dark,
      child: Scaffold(
        body: Listener(
          onPointerMove: (event) {
            var position = event.position.dy;
            var detal = position - prevDy;

            if (detal > 0) {
              // print("================向下移动================");
              debounce(() {
                setState(() {
                  _physics = ClampingScrollPhysics();
                });
              });
              var result = _scrollController.offset;
              if (result <= 0) {
                updatePicHeight(detal);
              }
            } else {
              // print("================向上移动================");
            }
          },
          onPointerUp: (_) {
            runAnimate();
            animationController.forward(from: 0.0);

            if (_scrollController.offset <= 0) {
              callNow = true;
              setState(() {
                _physics = AlwaysScrollableScrollPhysics();
              });
            }
          },
          onPointerDown: (event) {
            prevDy = event.position.dy;
            callNow = true;
            setState(() {
              _physics = AlwaysScrollableScrollPhysics();
            });
          },
          child: PrimaryScrollController(
            controller: _scrollController,
            child: CupertinoScrollbar(
              controller: _scrollController,
              child: CustomScrollView(
                physics: _physics,
                slivers: <Widget>[
                  SliverPersistentHeader(
                    pinned: true,
                    delegate: SliverCustomHeaderDelegate(
                        title: '账号',
                        collapsedHeight: 40,
                        expandedHeight: 300 + extraPicHeight,
                        paddingTop: MediaQuery.of(context).padding.top,
                        coverImgUrl:
                            'https://p1.music.126.net/fqdzslent-Jx2MmhWAh0ow==/109951163009168472.jpg'),
                  ),
                  SliverFixedExtentList(
                    itemExtent: 50.0,
                    delegate: new SliverChildBuilderDelegate(
                      (BuildContext context, int index) {
                        //创建列表项
                        return new Container(
                          alignment: Alignment.center,
                          color: Colors.lightBlue[100 * (index % 9)],
                          child: new Text('list item $index'),
                        );
                      },
                      childCount: 20, //50个列表项
                    ),
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}

class SliverCustomHeaderDelegate extends SliverPersistentHeaderDelegate {
  final double collapsedHeight;
  final double expandedHeight;
  final double paddingTop;
  final String coverImgUrl;
  final String title;
  String statusBarMode = 'dark';

  SliverCustomHeaderDelegate({
    this.collapsedHeight,
    this.expandedHeight,
    this.paddingTop,
    this.coverImgUrl,
    this.title,
  });

  @override
  double get minExtent => this.collapsedHeight + this.paddingTop;

  @override
  double get maxExtent => this.expandedHeight;

  @override
  bool shouldRebuild(SliverPersistentHeaderDelegate oldDelegate) {
    return true;
  }

  Color makeStickyHeaderBgColor(shrinkOffset) {
    if (shrinkOffset > 5) {
      return Colors.white;
    } else {
      return Colors.transparent;
    }
  }

  Color makeStickyHeaderTextColor(shrinkOffset) {
    if (shrinkOffset > 5) {
      return Color(0xFF000001);
    } else {
      return Colors.transparent;
    }
  }

  @override
  Widget build(
      BuildContext context, double shrinkOffset, bool overlapsContent) {
    return Container(
      height: this.maxExtent,
      width: MediaQuery.of(context).size.width,
      child: Stack(
        fit: StackFit.expand,
        children: <Widget>[
          Container(child: Image.network(this.coverImgUrl, fit: BoxFit.cover)),
          ClipRect(
            child: BackdropFilter(
              filter: ImageFilter.blur(sigmaX: 10.0, sigmaY: 10.0),
              child: Container(
                // height: 200.0,
                color: Colors.white.withOpacity(0.3),
              ),
            ),
          ),
          Positioned(
            left: 0,
            right: 0,
            top: 0,
            child: Container(
              color: this.makeStickyHeaderBgColor(shrinkOffset),
              child: SafeArea(
                bottom: false,
                child: Container(
                  height: this.collapsedHeight,
                  child: Row(
                    mainAxisAlignment: MainAxisAlignment.spaceBetween,
                    children: <Widget>[
                      IconButton(
                        icon: Icon(
                          IconData(0xe660, fontFamily: 'iconfont'),
                          color: Color(0xFF000001),
                        ),
                        onPressed: () {},
                      ),
                      Text(
                        this.title,
                        style: TextStyle(
                          fontSize: 20,
                          fontWeight: FontWeight.w500,
                          color: makeStickyHeaderTextColor(
                            shrinkOffset,
                          ),
                        ),
                      ),
                      IconButton(
                        icon: Icon(
                          IconData(0xe619, fontFamily: 'iconfont'),
                          color: Color(0xFF000001),
                        ),
                        onPressed: () {},
                      ),
                    ],
                  ),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

```
But this implementation is stupid.So how should I implement it using the `flexiblespacebar` and `stretchmode`.
I need to customize the Header Widget.I've been haunted by this problem for a long time.
Looking forward to your reply.

**Comment by Piinks** [comment]
Hi @1111mp, this may be a good question to post on stack overflow. Also, what version of flutter were you using when working with this sample code? I believe that this feature only arrived on the stable channel last week.

**Comment by 1111mp** [comment]
@Piinks Thanks for your replay.I use the latest version of stable channel(v1.9.1+hotfix.6).
I just learned about Flutter. There are some bugs in this sample code about elastic rolling.Flutter intends to support this function in the future?

**Comment by Piinks** [comment]
Thanks for the update! 

I am not sure what you mean by elastic rolling, but the stretch effect requested in this issue has been rolled out to the stable channel as of last week. 

If there is a bug you would like to report, I would recommend opening a new issue, with a description of the bug, the version of Flutter, and a minimal reproduction of the bug. 

From the sample code you have provided, it is hard to tell what the problem may be. It also looks like you are incorporating several custom widgets, as well as `ClampingScrollPhysics`, which do not support overscrolling. Feel free to ping me on the new issue so we can investigate this further. :)

**Comment by LOG-TAG** [comment]
![Image_zoom_example (1)](https://user-images.githubusercontent.com/1190344/78560899-6b2d8800-7834-11ea-811f-176d394a99b2.gif)

what is this animation called zoom in zoom out or stretchy zoomable header!

**Comment by diegoveloper** [comment]
@LOG-TAG I wrote a post about that : https://medium.com/@diegoveloper/flutter-sliverappbar-with-stretchy-header-9ca04f316ff0
